{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NCBI: National Center for Biotechnology Information\n",
    "#### Author: Ghassan Abboud - Genorobotics, Date : March 2024\n",
    "\n",
    "NCBI provides international genetic databases that serve as a basis for comparing the DNA fragments we obtain by sequencing. The database we use is GenBank, the result of a collaboration between the DNA Databank of Japan, the European Nucleotide Archive and NCBI. NCBI also provides a very efficient algorithm, BLAST, for aligning DNA sequences with their database and returning the best matches. \n",
    "\n",
    "Ways to access the GenBank database:\n",
    "- Search GenBank for sequences, names or identifiers: https://www.ncbi.nlm.nih.gov/genbank/\n",
    "- Using BLAST for sequence alignment compares your query to all entries in GenBank.\n",
    "- Use the Entrez Programming Utilities (see Part Two)\n",
    "\n",
    "## Part One: Exploring GenBank and BLASTn online\n",
    "\n",
    "Please refer to the pdf tutorial [_Exploring GenBank and BLASTn online_](Exploring_GenBank_and_BLASTn_online.pdf) on the repo.\n",
    "\n",
    "\n",
    "## Part Two: Accessing Genbank through Python\n",
    "\n",
    "### Introduction to E-Utilities\n",
    "\n",
    "NCBI provides a set of tools called the _Entrez Programming Utilities_ that give access to GenBank's query and database system through code. It allows you to retrieve information such as databases of genetic sequences or information on taxonomy. It uses a fixed URL syntax and HTTP requests, all starting with `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/`. Your main reference for coding pipelines that use these requests is the [Entrez Programming Utilities Help document](ncbi_entrez_eutilities_help.pdf) that provides examples in the PERL programming language. \n",
    "\n",
    "If you are comfortable with creating HTTP requests with Python, you can directly translate code from the Help document  and get inspiration from the sample applications on page 27. Creating the requests yourself probably provides more flexibility and control over the parameters. In this tutorial we will use a sublibrary of BioPython called `Biopython.Entrez` to create the requests. \n",
    "\n",
    "There are nine e-utilities, aka nine functions for accessing different information. the most important ones are:\n",
    "\n",
    "|E-utility|URL|description|parameters|\n",
    "|---|---|---|---|\n",
    "|ESearch (text searches)|eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi|Responds to a text query with the list of matching UIDs in a given database (for later use in ESummary, EFetch or ELink), along with the term translations of the query.| db = database, term = search term|\n",
    "|EFetch (data record downloads)|eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi|Responds to a list of UIDs in a given database with the corresponding data records in a specified format.| db = database, id = list of IDs separated by commas|\n",
    "|EPost (ID uploads)|eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi|Accepts a list of UIDs from a given database, stores the set on the History Server, and responds with a query key and web environment for the uploaded dataset.|db = database, id = list of IDs separated by commas|\n",
    "\n",
    "**Note:** to search on GenBank, the db parameter should be set equal to _nuccore_.\n",
    "\n",
    "**Note:** the list of IDs can be composed of GI or accession.version IDs. The term UID refers to any of these two.\n",
    "\n",
    "**Note:** the parameters listed are only the required ones. Before using a utility it is important to go over the optional parameters on page 39 because some like _retstart_ and _retmax_ come in handy.\n",
    "\n",
    "### Warning on NCBI requests frequency\n",
    "\n",
    "NCBI limits the number of requests that can be done by a single user to ensure the server is not overloaded. The limit is 3 requests per second. NCBI also recommends doing large jobs on weekends or between 9:00 PM and 5:00 AM Eastern time during weekdays If a user fails to comply with these guidelines, his IP address may get blocked.\n",
    "\n",
    "Hence, it is very important to properly investigate how the e-utilities can be used to execute a job in the _least amount of requests_. One should beware from using loops that contain e-utilities calls. One must also include their email address in any program they run, as we will see below. This way, NCBI may contact you in case of breach of guidelines instead of directly blocking your IP.\n",
    "\n",
    "### Biopython.Entrez\n",
    "\n",
    "this library provides functions that make the HTTP requests and return a handle to read the results. Handles are often used in Python to read ans write text files. You can checkout the library's [documentation](https://biopython.org/docs/1.75/api/Bio.Entrez.html).\n",
    "\n",
    "The functions in this module are named after their respective e-utilities and automatically respect the limit of 3 requests per second.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import os\n",
    "#First, one must set an email\n",
    "Entrez.email = 'fill.in.the.blanks@hotmail.com' #replace by your email address"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NM_005297               2115 bp    mRNA    linear   PRI 15-FEB-2024\n",
      "DEFINITION  Homo sapiens melanin concentrating hormone receptor 1 (MCHR1),\n",
      "            mRNA.\n",
      "ACCESSION   NM_005297\n",
      "VERSION     NM_005297.4\n",
      "KEYWORDS    RefSeq; MANE Select.\n",
      "SOURCE      Homo sapiens (human)\n",
      "  ORGANISM  Homo sapiens\n",
      "            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\n",
      "            Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;\n",
      "            Catarrhini; Hominidae; Homo.\n",
      "REFERENCE   1  (bases 1 to 2115)\n",
      "  AUTHORS   Takamura,N., Renaud,L., da Silveira,W.A. and Feghali-Bostwick,C.\n",
      "  TITLE     PDGF Promotes Dermal Fibroblast Activation via a Novel Mechanism\n",
      "            Mediated by Signaling Through MCHR1\n",
      "  JOURNAL   Front Immunol 12, 745308 (2021)\n",
      "   PUBMED   34912333\n",
      "  REMARK    GeneRIF: PDGF Promotes Dermal Fibroblast Activation via a Novel\n",
      "            Mechanism Mediated by Signaling Through MCHR1.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   2  (bases 1 to 2115)\n",
      "  AUTHORS   Vawter,M.P., Schulmann,A., Alhassen,L., Alhassen,W., Hamzeh,A.R.,\n",
      "            Sakr,J., Pauluk,L., Yoshimura,R., Wang,X., Dai,Q., Sanathara,N.,\n",
      "            Civelli,O. and Alachkar,A.\n",
      "  TITLE     Melanin Concentrating Hormone Signaling Deficits in Schizophrenia:\n",
      "            Association With Memory and Social Impairments and Abnormal\n",
      "            Sensorimotor Gating\n",
      "  JOURNAL   Int J Neuropsychopharmacol 23 (1), 53-65 (2020)\n",
      "   PUBMED   31563948\n",
      "  REMARK    GeneRIF: Melanin Concentrating Hormone Signaling Deficits in\n",
      "            Schizophrenia: Association With Memory and Social Impairments and\n",
      "            Abnormal Sensorimotor Gating.\n",
      "REFERENCE   3  (bases 1 to 2115)\n",
      "  AUTHORS   Badgandi,H.B., Hwang,S.H., Shimada,I.S., Loriot,E. and\n",
      "            Mukhopadhyay,S.\n",
      "  TITLE     Tubby family proteins are adapters for ciliary trafficking of\n",
      "            integral membrane proteins\n",
      "  JOURNAL   J Cell Biol 216 (3), 743-760 (2017)\n",
      "   PUBMED   28154160\n",
      "REFERENCE   4  (bases 1 to 2115)\n",
      "  AUTHORS   Fontaine-Bisson,B., Thorburn,J., Gregory,A., Zhang,H. and Sun,G.\n",
      "  TITLE     Melanin-concentrating hormone receptor 1 polymorphisms are\n",
      "            associated with components of energy balance in the Complex\n",
      "            Diseases in the Newfoundland Population: Environment and Genetics\n",
      "            (CODING) study\n",
      "  JOURNAL   Am J Clin Nutr 99 (2), 384-391 (2014)\n",
      "   PUBMED   24305679\n",
      "  REMARK    GeneRIF: Polymorphisms in the MCHR1 gene are associated with\n",
      "            differences in body composition and interact with physiologic and\n",
      "            energy-related lifestyle factors.\n",
      "REFERENCE   5  (bases 1 to 2115)\n",
      "  AUTHORS   Qu,C., Sun,X., Shi,Y., Gong,A., Liang,S., Zhao,M., Chen,Y. and\n",
      "            Liang,F.\n",
      "  TITLE     Microarray-based mutation detection of pediatric sporadic\n",
      "            nonsyndromic hearing loss in China\n",
      "  JOURNAL   Int J Pediatr Otorhinolaryngol 76 (2), 235-239 (2012)\n",
      "   PUBMED   22154049\n",
      "  REMARK    GeneRIF: SLC-1 protein human mutation demonstrated important\n",
      "            genetic susceptibility of sporadic nonsyndromic hearing loss in\n",
      "            China.\n",
      "REFERENCE   6  (bases 1 to 2115)\n",
      "  AUTHORS   Dunham,I., Shimizu,N., Roe,B.A., Chissoe,S., Hunt,A.R.,\n",
      "            Collins,J.E., Bruskiewich,R., Beare,D.M., Clamp,M., Smink,L.J.,\n",
      "            Ainscough,R., Almeida,J.P., Babbage,A., Bagguley,C., Bailey,J.,\n",
      "            Barlow,K., Bates,K.N., Beasley,O., Bird,C.P., Blakey,S.,\n",
      "            Bridgeman,A.M., Buck,D., Burgess,J., Burrill,W.D., O'Brien,K.P. et\n",
      "            al.\n",
      "  TITLE     The DNA sequence of human chromosome 22\n",
      "  JOURNAL   Nature 402 (6761), 489-495 (1999)\n",
      "   PUBMED   10591208\n",
      "  REMARK    Erratum:[Nature 2000 Apr 20;404(6780):904]\n",
      "REFERENCE   7  (bases 1 to 2115)\n",
      "  AUTHORS   Shimomura,Y., Mori,M., Sugo,T., Ishibashi,Y., Abe,M., Kurokawa,T.,\n",
      "            Onda,H., Nishimura,O., Sumino,Y. and Fujino,M.\n",
      "  TITLE     Isolation and identification of melanin-concentrating hormone as\n",
      "            the endogenous ligand of the SLC-1 receptor\n",
      "  JOURNAL   Biochem Biophys Res Commun 261 (3), 622-626 (1999)\n",
      "   PUBMED   10441476\n",
      "REFERENCE   8  (bases 1 to 2115)\n",
      "  AUTHORS   Saito,Y., Nothacker,H.P., Wang,Z., Lin,S.H., Leslie,F. and\n",
      "            Civelli,O.\n",
      "  TITLE     Molecular characterization of the melanin-concentrating-hormone\n",
      "            receptor\n",
      "  JOURNAL   Nature 400 (6741), 265-269 (1999)\n",
      "   PUBMED   10421368\n",
      "REFERENCE   9  (bases 1 to 2115)\n",
      "  AUTHORS   Chambers,J., Ames,R.S., Bergsma,D., Muir,A., Fitzgerald,L.R.,\n",
      "            Hervieu,G., Dytko,G.M., Foley,J.J., Martin,J., Liu,W.S., Park,J.,\n",
      "            Ellis,C., Ganguly,S., Konchar,S., Cluderay,J., Leslie,R., Wilson,S.\n",
      "            and Sarau,H.M.\n",
      "  TITLE     Melanin-concentrating hormone is the cognate ligand for the orphan\n",
      "            G-protein-coupled receptor SLC-1\n",
      "  JOURNAL   Nature 400 (6741), 261-265 (1999)\n",
      "   PUBMED   10421367\n",
      "REFERENCE   10 (bases 1 to 2115)\n",
      "  AUTHORS   Kolakowski,L.F. Jr., Jung,B.P., Nguyen,T., Johnson,M.P.,\n",
      "            Lynch,K.R., Cheng,R., Heng,H.H., George,S.R. and O'Dowd,B.F.\n",
      "  TITLE     Characterization of a human gene related to genes encoding\n",
      "            somatostatin receptors\n",
      "  JOURNAL   FEBS Lett 398 (2-3), 253-258 (1996)\n",
      "   PUBMED   8977118\n",
      "COMMENT     REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "            reference sequence was derived from Z86090.10.\n",
      "            This sequence is a reference standard in the RefSeqGene project.\n",
      "            \n",
      "            On Jun 26, 2020 this sequence version replaced NM_005297.3.\n",
      "            \n",
      "            Summary: The protein encoded by this gene, a member of the G\n",
      "            protein-coupled receptor family 1, is an integral plasma membrane\n",
      "            protein which binds melanin-concentrating hormone. The encoded\n",
      "            protein can inhibit cAMP accumulation and stimulate intracellular\n",
      "            calcium flux, and is probably involved in the neuronal regulation\n",
      "            of food consumption. Although structurally similar to somatostatin\n",
      "            receptors, this protein does not seem to bind somatostatin.\n",
      "            [provided by RefSeq, Jul 2008].\n",
      "            \n",
      "            Publication Note:  This RefSeq record includes a subset of the\n",
      "            publications that are available for this gene. Please see the Gene\n",
      "            record to access additional publications.\n",
      "            \n",
      "            ##Evidence-Data-START##\n",
      "            Transcript exon combination :: AF490537.2, BC021146.1 [ECO:0000332]\n",
      "            RNAseq introns              :: single sample supports all introns\n",
      "                                           SAMEA1965299, SAMEA1966682\n",
      "                                           [ECO:0000348]\n",
      "            ##Evidence-Data-END##\n",
      "            \n",
      "            ##RefSeq-Attributes-START##\n",
      "            MANE Ensembl match     :: ENST00000249016.5/ ENSP00000249016.5\n",
      "            RefSeq Select criteria :: based on single protein-coding transcript\n",
      "            ##RefSeq-Attributes-END##\n",
      "            COMPLETENESS: complete on the 3' end.\n",
      "PRIMARY     REFSEQ_SPAN         PRIMARY_IDENTIFIER PRIMARY_SPAN        COMP\n",
      "            1-251               Z86090.10          100158-100408\n",
      "            252-2115            Z86090.10          101623-103486\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..2115\n",
      "                     /organism=\"Homo sapiens\"\n",
      "                     /mol_type=\"mRNA\"\n",
      "                     /db_xref=\"taxon:9606\"\n",
      "                     /chromosome=\"22\"\n",
      "                     /map=\"22q13.2\"\n",
      "     gene            1..2115\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"melanin concentrating hormone receptor 1\"\n",
      "                     /db_xref=\"GeneID:2847\"\n",
      "                     /db_xref=\"HGNC:HGNC:4479\"\n",
      "                     /db_xref=\"MIM:601751\"\n",
      "     exon            1..251\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /inference=\"alignment:Splign:2.1.0\"\n",
      "     CDS             170..1231\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"G protein-coupled receptor 24; MCH receptor 1;\n",
      "                     somatostatin receptor-like protein\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"melanin-concentrating hormone receptor 1\"\n",
      "                     /protein_id=\"NP_005288.4\"\n",
      "                     /db_xref=\"CCDS:CCDS14004.2\"\n",
      "                     /db_xref=\"GeneID:2847\"\n",
      "                     /db_xref=\"HGNC:HGNC:4479\"\n",
      "                     /db_xref=\"MIM:601751\"\n",
      "                     /translation=\"MDLEASLLPTGPNASNTSDGPDNLTSAGSPPRTGSISYINIIMP\n",
      "                     SVFGTICLLGIIGNSTVIFAVVKKSKLHWCNNVPDIFIINLSVVDLLFLLGMPFMIHQ\n",
      "                     LMGNGVWHFGETMCTLITAMDANSQFTSTYILTAMAIDRYLATVHPISSTKFRKPSVA\n",
      "                     TLVICLLWALSFISITPVWLYARLIPFPGGAVGCGIRLPNPDTDLYWFTLYQFFLAFA\n",
      "                     LPFVVITAAYVRILQRMTSSVAPASQRSIRLRTKRVTRTAIAICLVFFVCWAPYYVLQ\n",
      "                     LTQLSISRPTLTFVYLYNAAISLGYANSCLNPFVYIVLCETFRKRLVLSVKPAAQGQL\n",
      "                     RAVSNAQTADEERTESKGT\"\n",
      "     misc_feature    206..208\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"N-linked (GlcNAc...) asparagine.\n",
      "                     /evidence=ECO:0000255; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q99705.3); glycosylation site\"\n",
      "     misc_feature    215..217\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"N-linked (GlcNAc...) asparagine.\n",
      "                     /evidence=ECO:0000255; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q99705.3); glycosylation site\"\n",
      "     misc_feature    236..238\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"N-linked (GlcNAc...) asparagine.\n",
      "                     /evidence=ECO:0000255; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q99705.3); glycosylation site\"\n",
      "     misc_feature    302..370\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     misc_feature    407..475\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     misc_feature    518..586\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     misc_feature    644..712\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     misc_feature    797..865\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     misc_feature    926..994\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     misc_feature    1037..1105\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q99705.3);\n",
      "                     transmembrane region\"\n",
      "     exon            252..2115\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /inference=\"alignment:Splign:2.1.0\"\n",
      "     regulatory      2084..2089\n",
      "                     /regulatory_class=\"polyA_signal_sequence\"\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"hexamer: AATAAA\"\n",
      "     polyA_site      2115\n",
      "                     /gene=\"MCHR1\"\n",
      "                     /gene_synonym=\"GPR24; MCH-1R; MCH1R; SLC-1; SLC1\"\n",
      "                     /note=\"major polyA site\"\n",
      "ORIGIN      \n",
      "        1 agttgggctt ggaggcggca gcggctgcca ggctacggag gaagaccccc ttcccaactg\n",
      "       61 cggggcttgc gctccgggac aaggtggcag gcgctggagg ctgccgcagc ctgcgtgggt\n",
      "      121 ggaggggagc tcagctcggt tgtgggagca ggcgaccggc actggctgga tggacctgga\n",
      "      181 agcctcgctg ctgcccactg gtcccaacgc cagcaacacc tctgatggcc ccgataacct\n",
      "      241 cacttcggca ggatcacctc ctcgcacggg gagcatctcc tacatcaaca tcatcatgcc\n",
      "      301 ttcggtgttc ggcaccatct gcctcctggg catcatcggg aactccacgg tcatcttcgc\n",
      "      361 ggtcgtgaag aagtccaagc tgcactggtg caacaacgtc cccgacatct tcatcatcaa\n",
      "      421 cctctcggta gtagatctcc tctttctcct gggcatgccc ttcatgatcc accagctcat\n",
      "      481 gggcaatggg gtgtggcact ttggggagac catgtgcacc ctcatcacgg ccatggatgc\n",
      "      541 caatagtcag ttcaccagca cctacatcct gaccgccatg gccattgacc gctacctggc\n",
      "      601 cactgtccac cccatctctt ccacgaagtt ccggaagccc tctgtggcca ccctggtgat\n",
      "      661 ctgcctcctg tgggccctct ccttcatcag catcacccct gtgtggctgt atgccagact\n",
      "      721 catccccttc ccaggaggtg cagtgggctg cggcatacgc ctgcccaacc cagacactga\n",
      "      781 cctctactgg ttcaccctgt accagttttt cctggccttt gccctgcctt ttgtggtcat\n",
      "      841 cacagccgca tacgtgagga tcctgcagcg catgacgtcc tcagtggccc ccgcctccca\n",
      "      901 gcgcagcatc cggctgcgga caaagagggt gacccgcaca gccatcgcca tctgtctggt\n",
      "      961 cttctttgtg tgctgggcac cctactatgt gctacagctg acccagttgt ccatcagccg\n",
      "     1021 cccgaccctc acctttgtct acttatacaa tgcggccatc agcttgggct atgccaacag\n",
      "     1081 ctgcctcaac ccctttgtgt acatcgtgct ctgtgagacg ttccgcaaac gcttggtcct\n",
      "     1141 gtcggtgaag cctgcagccc aggggcagct tcgcgctgtc agcaacgctc agacggctga\n",
      "     1201 cgaggagagg acagaaagca aaggcacctg atacttcccc tgccaccctg cacacctcca\n",
      "     1261 agtcagggca ccacaacacg ccaccgggag agatgctgag aaaaacccaa gaccgctcgg\n",
      "     1321 gaaatgcagg aaggccgggt tgtgaggggt tgttgcaatg aaataaatac attccatggg\n",
      "     1381 gctcacacgt tgctggggag gcctggagtc aggtttgggg ttttcagata tcagaaatcc\n",
      "     1441 ccttggggga gcaggatgag acctttggat agaacagaag ctgagcaaga gaacatgttg\n",
      "     1501 gtttggataa ccggttgcac tatatctgtg agctctcaaa tgtcttcttc ccaaggcaag\n",
      "     1561 aggtggaagg gtactgactg ggtttgttta aagtcaggca gggctggagt gagcagccag\n",
      "     1621 ggccatgttg cacaaggcct gagagacggg aaagggcccg atcgctcttt cccgcctctc\n",
      "     1681 actggtgcga tggaaggtgg cctttctccc aagctggtgg ataatgaaaa ataaagcatc\n",
      "     1741 ccatctctcg gcgttccagc atcctgtcaa tttccctttt gctctagagg atgcatgttt\n",
      "     1801 atttgagggg atgtggcact gagcccacag gagtaaaagc ccagtttgct aggaggtctg\n",
      "     1861 cttactgaaa acaaggagac ctggggtggg tgtggttggg ggtcttaaaa ctaataaaag\n",
      "     1921 ctggggtcgg ggggcttttg cagctctggt gacattctct ccacggggca catttgctca\n",
      "     1981 gtcactaatc cagcttgagt gtccgtgtgt tctgcatgtg caggggtcat tctagtgccc\n",
      "     2041 ggtgtgttgg catcatcttt ttgctctagc ccttcctctc caaaataaaa tcaaataaag\n",
      "     2101 gaaaatctcc accca\n",
      "//\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Let's first do a simple request\n",
    "handle = Entrez.efetch(db = \"nucleotide\", id = \"NM_005297.4\", rettype = \"gb\", retmode = \"text\")\n",
    "#This line launches a request that donwloads the entry corresponding to ID NM_005297.4\n",
    "#This parameter rettype and retmode are explained in the Eutilities Help book\n",
    "entry = handle.read()\n",
    "handle.close()\n",
    "print(entry)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Let's now download this record again as a fasta file and writing it locally\n",
    "handle = Entrez.efetch(db = \"nucleotide\", id = \"NM_005297.4\", rettype = \"fasta\", retmode = \"fasta\")\n",
    "sequence = handle.read()\n",
    "handle.close()\n",
    "path = os.getcwd()+ \"\\extracted_sequence.fasta\" #get current working directory and add file name\n",
    "with open(path, mode = \"w\") as files: #we open a handle into a new file in our current working directory\n",
    "   files.write(sequence) #and we write the sequence inside"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After running the above cell you should see a new file appear with your extracted sequence. For the next level, let's combine efetch with esearch.\n",
    "We will perform a search using a term as if we were on the GenBank website. This search will return IDs that we will then plug into efetch to retrieve the associated GenBank entries.\n",
    "\n",
    "By default, `Entrez.esearch` only returns the top 20 results. We can change that with the optional keyword `retmax`. Note that the maximum value of retmax is 10,000. For searches returning more than 10,000 results, they must be downloaded in batches.\n",
    "\n",
    "Compared to efetch, esearch always returns the results in XML format, which is not very handy, we can parse it into a Python dictionary with `Entrez.read`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PP109813.1', 'OR475126.1', 'OR475125.1', 'OR475124.1', 'OR475123.1', 'OR475122.1', 'OR475121.1', 'OR475120.1', 'OR475119.1', 'OR475118.1', 'OR475117.1', 'OR475116.1', 'OR475115.1', 'OR475114.1', 'OR475113.1', 'OR475112.1', 'OR475111.1', 'OR475110.1', 'OR475109.1', 'OR475108.1', 'OR475107.1', 'OR475106.1', 'OR475105.1', 'OR475104.1', 'OR475103.1', 'OR475102.1', 'OR475101.1', 'NZ_JAMQON010000001.1', 'NZ_CP010835.1', 'NZ_CP013050.1', 'NZ_CP009149.1', 'NC_018015.1', 'NC_013922.1', 'LC777626.1', 'OR885584.1', 'OR936736.1', 'OR936735.1', 'OR936734.1', 'OR936733.1', 'OR936732.1', 'PP390563.1', 'PP391058.1', 'PP391057.1', 'PP383999.1', 'PP383998.1', 'PP383997.1', 'PP383996.1', 'PP383995.1', 'PP383994.1', 'PP383993.1']\n"
     ]
    }
   ],
   "source": [
    "search_term = \"rbcl\" #we search for entries corresponding to the rbcl gene\n",
    "handle = Entrez.esearch(db = \"nucleotide\", term = search_term, retmax = 50, idtype = \"acc\")\n",
    "search_result = Entrez.read(handle)\n",
    "handle.close()\n",
    "relevant_ids= search_result[\"IdList\"]\n",
    "print(relevant_ids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.efetch(db=\"nucleotide\", id=relevant_ids, retmode = \"fasta\", rettype = \"fasta\")\n",
    "rbcl_sequences = handle.read()\n",
    "handle.close()\n",
    "path = os.getcwd()+ r\"\\\\rbcl_database.fasta\" #get current working directory and add file name\n",
    "with open(path, mode = \"w\") as file: #we open a handle into a new file in our current working directory\n",
    "   file.write(rbcl_sequences) #and we write the sequence inside\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the last example, we downloaded a database of the 20 first entries. Look at the first one in your downloaded file, it is the **complete genome** of a plant. To avoid such situations, we will now learn to filter our results by sequence length and other parameters, like we did on the webpage by using advanced search.\n",
    "\n",
    "The e-Utilities do not take optional arguments like sequence length, _all these parameters need to be included in the search term, following the same structure as the one built by the online advanced search tool (see Part One)_. Here is a recommended workflow for developing these search terms:\n",
    "- Go on the GenBank website and use [Advanced Search](https://www.ncbi.nlm.nih.gov/nuccore/advanced)\n",
    "- fill in the fields you want to filter by\n",
    "- copy the term from the search bar and use it in your code.\n",
    "\n",
    "This seems counter-intuitive, why even bother downloading it through code if I have to go on the website anyway? Yes! However, by understanding how the search term is built, one can code a function that takes as arguments some fields and builds the search term, as in the example below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_gene_database(gene_name, min_sequence_length, max_sequence_length, file_name, retmax):\n",
    "    '''Download data to a .fasta file, with filtering options\n",
    "\n",
    "    Args:\n",
    "        gene_name: (str) name of the gene to extract from GenBank\n",
    "        min_sequence_length: (int) lower bound of the sequence length filter\n",
    "        max_sequence_length: (int) upper bound of the sequence length filter\n",
    "        file_name:(str) name of the generated file\n",
    "        retmax:(int) number of entries to be downloaded, smaller than 10 000\n",
    "    '''\n",
    "    #building search term\n",
    "    search_term = f\"{gene_name}[Gene Name] AND {min_sequence_length}:{max_sequence_length}[Sequence Length]\"\n",
    "\n",
    "    #searching for the IDs of the entries that match the search\n",
    "    handle = Entrez.esearch(db = \"nucleotide\", term = search_term, retmax = retmax, idtype = \"acc\")\n",
    "    search_result = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    relevant_ids= search_result[\"IdList\"]\n",
    "\n",
    "    #Using these IDs to retrieve the entries\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=relevant_ids, retmode = \"fasta\", rettype = \"fasta\")\n",
    "    sequences = handle.read()\n",
    "    handle.close()\n",
    "\n",
    "    #writing the entries to our own local FASTA file\n",
    "    path = os.getcwd()+ r\"\\\\\"+ file_name\n",
    "    with open(path, mode = \"w\") as file:\n",
    "        file.write(sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "download_gene_database(\"matk\", 750, 1500, \"my_gene_database.fasta\", 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Bonus: can you edit the function's body to be able to donwload more than 10 000 entries.\n",
    "#Reminder: one esearch call cannot download more than 10000 entries even if retmax is set higher\n",
    "#Hint: check the help book for the retstart optional parameter"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
